Initial attempt at weight Breakdown (issue #507)

mayork opened the pull request:
@whoburg initial breakdown class ready for review. Should be fully functional. Let me know if you want to tweak how the diagram looks, I could possibly make it colored, change fonts/sizes, dash some lines, etc.

current problems:
improper indent warning from pylint on continued lines

acdl-jenkins commented:
Can one of the admins verify this patch?

whoburg commented:
test this please

whoburg commented:
@mayork, I'm going to rebase and force push to eliminate all the commits related to a merge gone wrong. You'll need to update your local branch -- I'll post again with instructions on how to do that

whoburg commented:
ok, rebased and force pushed (and I have a backup of the branch in case there is any issue)

anyone (including @mayork ) that has a local checkout of this branch needs to do the following:

(stash any uncommitted changes)
```
git checkout master
git branch -D breakdown
git checkout breakdown
git pull origin breakdown
```

whoburg commented:
@mayork, it looks like your text editor is still set up to insert tabs instead of 4 spaces -- google how to set the `tabstop` and `expandtab` so that tab means 4 spaces for python files.

whoburg commented on gpkit/breakdown.py:
```diff
@@ -0,0 +1,92 @@
+        self.constr = []
+        self.varlist = []
+        #call recursive function to create gp constraints
+        total = self.recurse(input_dict)
```
`total` is never used -- should it be?

whoburg commented on gpkit/breakdown.py:
```diff
@@ -0,0 +1,92 @@
+
+    def solve_method(self):
+        m = Model(self.make_objective(), self.make_constraints())
+        return m.solve(verbosity=0)
```
I don't think this method is needed. no need to solve the model on initialization -- let a user cal `solve()` instead.

whoburg commented on gpkit/breakdown.py:
```diff
@@ -0,0 +1,92 @@
+        self.sol = self.solve_method()
+
+    def getSolution(self):
+        return self.sol
```
no need for this either -- see below

whoburg commented on gpkit/breakdown.py:
```diff
@@ -0,0 +1,92 @@
+        m = Model(self.make_objective(), self.make_constraints())
+        return m.solve(verbosity=0)
+
+    def recurse(self, input_dict):
```
need docstring

mayork replied:
@whoburg this is the default variable description?

whoburg commented on gpkit/breakdown.py:
```diff
@@ -0,0 +1,92 @@
+            while j < len(self.constr[i][1]):
+                posy = posy+self.constr[i][1][j]
+                j = j+1
+            constraints.append(1 >= posy/self.constr[i][0])
```
might be clearer to do `constraints.append(self.constr[i][0] >= posy)` ?

mayork replied:
that makes sense as well...so we don't need to have everything in "standard" format (I think in one of Boyd's lectures he said standard format was to have a one on one side)?

whoburg replied:
nope, that conversion is handled internally by gpkit

whoburg commented on gpkit/tests/t_model.py:
```diff
@@ -354,6 +355,23 @@ def test_signomials_not_allowed_in_objective(self):
             with self.assertRaises(TypeError):
                 sol = m.localsolve(verbosity=0)
 
+    def test_breakdown(self):
```
this should not be inside the `TestSP` class -- make a separate `TestBreakdown` class, separate from `TestGP` and `TestSP`

whoburg commented:
@mayork I did an initial code review with inline comments. You can add additional commits to address them. I think there are some ways to further simplify this - but you can start with the comments so far.

mayork commented:
disregard the large chunk of deleted lines in the above commits...I was getting a merge error when I tried to commit, I think this was an artifact of moving the branch around. It's fixed now.

mayork commented:
@whoburg addressed all of the corrections and am up and running with pylint

mayork commented:
@whoburg talked with @bqpd about the graphics...he thought SVG was the way to go. If you concur I'll start on that.

acdl-jenkins commented:
Can one of the admins verify this patch?

mayork commented:
@whoburg 
![image](https://cloud.githubusercontent.com/assets/6796415/13652901/d7d9b9f0-e61c-11e5-846d-d92949ba90d6.png)

this is my current diagram...thoughts?

acdl-jenkins commented:
Can one of the admins verify this patch?

bqpd commented:
test this please

bqpd commented:
retest this please

bqpd commented:
verify this please (is that how I do this)

bqpd commented:
Looks like there's merge conflicts?

mayork commented:
@bqpd not sure how to fix a merge conflict...also, what I just committed doesn't pass pylint because of improper indent on a continued line. What is the protocol for that? I couldn't figure it our from the config file

whoburg commented:
retest this please

whoburg commented:
@mayork can you update the very first comment to give an overview of the current status? What's there now is out of date.

whoburg commented:
ok failing unit tests because `svgwrite` is not on the Jenkins machines. @bqpd any strategy thoughts here? Should part of this be moved to interactive?

bqpd commented:
err, is `breakdown.py` the only changed code?

bqpd commented:
I think this should inherit from `ConstraintSet` and have a method which imports the svg drawing logic from interactive, like `Model` and `ControlPanel`.

whoburg commented:
retest this please

bqpd commented on gpkit/breakdown.py:
```diff
@@ -188,16 +188,16 @@ def test(self):
+        npt.assert_almost_equal(self.sol('w121').magnitude, 2, decimal=5)
+        npt.assert_almost_equal(self.sol('w122').magnitude, 6, decimal=5)
+        npt.assert_almost_equal(self.sol('w21').magnitude, 1, decimal=5)
+        npt.assert_almost_equal(self.sol('w22').magnitude, 2, decimal=5)
```
`mag(sol(...))` will work even if units are disabled.

mayork commented:
@bqpd @whoburg I still am not sure what exactly needs to be moved to svg.py inside of interactive....do I need to move every method that touches any part of the drawing object (i.e. dwgrecurese and drawsegment)? I have tried a few things and can't something which works

when I try and do this:

from .interactive.svg import importsvgwrite
        from svgwrite import cm

i get this error
"ValueError: Attempted relative import in non-package"

I just pushed with the svg.py file I made inside interactive and an updated breakdown.py...the svg.py has some commented out areas which reflect a previous attempt to get things working

whoburg commented:
Try `from gpkit.interactive.svg import` ...

I don't fully understand why the relative import isn't working here - @bqpd might know.

In any case, try to follow the pattern currently used by `gpkit/constraints/costed`'s `controlpanel` method, which imports `modelcontrolpanel` from interactive.

bqpd commented:
I think the problem is that this isn't being imported as part of the module, and so the relative import won't work.

mayork commented:
@bqpd could we meet to look at this? I tried a few other things...if I try to do from gpkit.interactive.svg import _____ i get the error:

importerror: no module named svg

bqpd commented:
I think for now you should just import svgwrite in `breakdown.py`; when the breakdown class is feature-complete, I can sit down with you and we can figure out how to integrate it into gpkit.

mayork commented:
@bqpd I think that's the point we're at, woody was hoping to merge it into master and close this branch out this week/next week

bqpd commented:
Okay! For this to be in GPkit, Breakdown should probably be a ConstraintSet which generates its constraints on init, and its name should end with ConstraintSet. Then should be a function in interactive which accepts a Breakdown and does all the plotting; much of what are attributes of `self` now will be instead gotten from the passed Breakdown.

bqpd commented:
I'll take a stab at starting this.

bqpd commented:
Ready for review by @mayork

mayork commented:
@bqpd @whoburg I think this is ready for final review...I moved all the unit tests to t_constraints, all unit tests pass on my machine. Also fixed some diagram drawing problems that came up after moving things to interactive

not sure where the branch conflicts are...i can't find them

bqpd commented on gpkit/constraints/breakdown.py:
```diff
@@ -31,8 +31,17 @@ def __init__(self, input_dict, units):
+        self.constraints = constraints
         ConstraintSet.__init__(self, constraints)
 
+    def solve_method(self):
```
What's wrong with Breakdown just being a constraintset?

bqpd commented on gpkit/tests/t_constraints.py:
```diff
@@ -125,8 +126,31 @@ def test_init(self):
 
-
-TESTS = [TestConstraint, TestMonomialEquality, TestSignomialInequality]
+class TestBreakdown(unittest.TestCase):
```
Niiice

bqpd replied:
Unfortunately putting the test here leads to it being tested once for every installed solver...

bqpd commented on bd.py:
```diff
@@ -3,11 +3,13 @@
 BD = Breakdown(TEST, "N")
 m = Model(BD.varlist[0], BD)
-sol = m.solve()
+sol = BD.solve_method()#m.solve()
```
m is now unused.

bqpd commented:
@mayork the conflicts are probably from some PRs recently merged to master, I can handle that. Breakdown still doesn't match the gpkit code style in some significant ways; let's meet up in person and go over the code before final review by @whoburg.

mayork commented:
@bqpd sounds good, when would you like to meet?

bqpd commented:
retest this please

bqpd commented:
Ready to merge once TODOs are cleaned up and `bd.py` is removed. There should potentially be a test that the correct SVG is produced?

mayork commented:
@whoburg @bqpd took care of all of the TODOs, just need to remove bd.py

Also, what kind of test did you want to verify the correct SVG is produced? Should I just save a sample file into gpkit somewhere?

bqpd commented on gpkit/constraints/breakdown.py:
```diff
@@ -43,11 +46,8 @@ def _recurse(self, input_dict, units):
 
     def make_diagram(self, sol, sidelength=12, height=20):
```
I think filename should be an argument here, not in `__init__`

bqpd commented:
Awesome! Reviewing...

bqpd commented on gpkit/interactive/svg.py:
```diff
@@ -4,85 +4,74 @@
-    while i < len(order):
-        height = int(round((((bd.height/bd.total)*sol(order[i])))))
-        name = order[i]
+    for key in sorted(input_dict):
```
if you use `for key, value in sorted(input_dict.items()):` you should be able to replace `input_dict[key]` below with `value`.

mayork replied:
Yeah I think the most current version already has that...is it possible you're looking at an outdated diff?

bqpd commented on gpkit/interactive/svg.py:
```diff
@@ -0,0 +1,78 @@
+        currentcoord = (initcoord[0], initcoord[1]+totalheight)
+        drawsegment(key, height, currentcoord, dwg, elementlength)
+        totalheight = totalheight+height
+        if isinstance(input_dict[key], dict):
```
@mayork, this still seems to be in the latest changes. Perhaps you've a local commit you haven't pushed yet?

mayork replied:
@bqpd oh wow sorry I'm an idiot and totally misunderstood what the first comment was about. For some reason my git is saying you'r commenting on an outdated diff...but best as I can tell this is the most current thing there is, it seems the same as what I have locally

bqpd replied:
No worries!

mayork replied:
@bqpd  just made line 36 of svg.py read this on my local copy

if isinstance(value, dict):

and I get this error: NameError: global name 'value' is not defined

I just tried to replace input_dict[key] with value...what did i do wrong?

bqpd replied:
Line 31 should be `for key, value in sorted(input_dict.items()):`

mayork commented:
@bqpd just pushed a new version with your changes

mayork commented:
@whoburg I looked over this again and think it's ready for review...do you want any sort of test method? I'm not entirely sure what that would/should look like

whoburg commented:
just rebased this with @mayork, force pushing

whoburg commented:
retest this please

mayork commented:
<img width="627" alt="screen shot 2016-08-30 at 2 39 06 pm" src="https://cloud.githubusercontent.com/assets/6796415/18102178/a676f806-6ebf-11e6-8434-e9812a4fbf8a.png">
this is the diagram the test method draws...seems potentially useful to have somewehre

bqpd commented:
Nice, this looks good.

mayork commented:
@whoburg this is ready for a final review. I think we should disable the remaining pylint warnings, but I didn't want to do that before it was reviewed

whoburg commented:
retest this please

whoburg commented:
@mayork I'd like to do  this review in two stages:
1. high-level comments/discussion (now)
2. a more detailed line-by-line review by me or @bqpd

For now, the high-level comments/discussion:

1. I want us to try using this on the 1682 model and see how it goes. Could you please work with @mjburton11 to try this out on a gpkit-models branch. My suspicion is that it won't make sense to specify a single weight breakdown input dict in one location -- rather, it will be more natural to have multiple Breakdown objects (wing, fuselage, etc) that are local to their submodels/components -- but not sure how this will really play out.

2. Related to the above, does it work for one of the leaves of Breakdown A to itself be a Breakdown?

3. Further related to the above, should we implement it that way, i.e. each Breakdown is only one level deep, its leaves always being either a Variable or a Breakdown?

4. Can we switch the font to math font? Have you tested a latex math symbol like `\rho`?

(@bqpd, any comments?)

mayork commented:
1. I'll get with Mike and set up a 1682 breakdown. 

2. Have not tested that...however, if it doesn't work should be an easy feature to add simply by appending the constraints from the other breakdown

3. What do you see as the benefit of that method? I'm not sure I fully understand what you're going for 

4. Changing font is possible...I looked into it before, it was a while ago so I don't recall why I didn't do it. I'm not sure on math type, I'm guessing there's a way to do it but I need to investigate

whoburg commented:
The benefit of 3 would be a cleaner implementation in that the data structures and parsing would match up. To parse an input dict, for all values in a level of the dict you would a) create a variable, if the value is a string or Variable, or b) create a Breakdown, if the value is a dict.

Additional benefit is that it would be always possible to draw the breakdown diagram associated with a subset of a Breakdown, e.g. Aircraft has a Breakdown, but its Wing is also a breakdown.

Start by addressing 1 and I suspect 3 may end up being related.

mayork commented:
@whoburg more detailed answer to 4 from above. new fonts and styles can be defined as a new CSS style...I'm not venturing outside my knowledge base but from what I know/can tell from some quick web research CSS has limited support for math, so I don't think we can go display math font. How big of a problem is that? I think it would be nice to be able to do, but I don't think it's that big of an issue

whoburg commented:
I would be surprised if the svg package does not support latex math. But I've been surprised before.

In other news, looks like we have some conflicts; I'll attempt a rebase.

mayork commented:
<img width="262" alt="screen shot 2016-09-06 at 8 54 56 pm" src="https://cloud.githubusercontent.com/assets/6796415/18296013/4dceca70-7474-11e6-8cc1-530dbf1362bd.png">
<img width="252" alt="screen shot 2016-09-06 at 8 55 09 pm" src="https://cloud.githubusercontent.com/assets/6796415/18296017/51085012-7474-11e6-9780-abba12adaccc.png">
<img width="242" alt="screen shot 2016-09-06 at 8 55 33 pm" src="https://cloud.githubusercontent.com/assets/6796415/18296018/52c9890c-7474-11e6-81bb-ceccc2ea6be7.png">
<img width="371" alt="screen shot 2016-09-06 at 8 55 19 pm" src="https://cloud.githubusercontent.com/assets/6796415/18296019/5480b572-7474-11e6-9fd4-d745f4c5b593.png">

@mjburton11 @whoburg 16.82 gas_male breakdowns. the code to make them is below.

```python
    #create the centbreakdown
    BDcent = {'Center': {'Fuel': sol('W_{fuel-tot}'), 'Payload': sol('W_{pay}'),
                                'Engine': sol('W_{eng-tot}'), 'Avionics': sol('W_{avionics}'),
                                'Skid': sol('W_{skid}')}}
    centbd = Breakdown(BDcent, "lbf")
    mcent = Model(centbd.root, centbd)
    centsol = mcent.solve()
    centbd.make_diagram(centsol,  "16.82_cent_weight_breakdown.svg", 7, 20)

    #create the fuselage breakdown
    BDfuse = {'Fuselage': {'Ribs': [sol('m_{rib}')*9.8*units('m/s^2'), "N"],
                           'Skin': [sol('m_{fuse}')*9.8*units('m/s^2'), "N"]}}
    fusebd = Breakdown(BDfuse, "lbf")
    mfuse = Model(fusebd.root, fusebd)
    fusesol = mfuse.solve()
    fusebd.make_diagram(fusesol, "16.82_fuse_weight_breakdown.svg", 7, 12)

    #create the wing breakdown
    BDwing = {'Wing': {'Caps': [sol('m_{cap}')*9.8*units('m/s^2'), "N"],
                       'Skin':[sol('m_{skin}')*9.8*units('m/s^2'), "N"]}}
    wingbd = Breakdown(BDwing, "lbf")
    mwing = Model(wingbd.root, wingbd)
    wingsol = mwing.solve()
    wingbd.make_diagram(wingsol, "16.82_wing_weight_breakdown.svg", 7, 12)

    #create the total breakdown
    total = dict(BDwing.items() + BDfuse.items() + BDcent.items())
    BDtotal = {'Total': total}
    fullbd = Breakdown(BDtotal, "lbf")
    mfull = Model(fullbd.root, fullbd)
    bdsol = mfull.solve(solver="mosek")
    fullbd.make_diagram(bdsol, "16.82_weight_breakdown.svg", 10, 22)
```

mayork commented:
@whoburg you cannot give a Breakdown another Breakdown as an arugment. However, I think the way it currently works is still close to your vision.

Breakdown is really just a plotting tool. The user supplies a Breakdown with a nested input dict which defines the breakdown. Breakdown provides a convenient method to sum weights in this input dict, but that does not need to be used if the values for all the weights are already known. If the user wants multiple breakdowns like I made above, you can just make multiple sub dicts and combine them before making a Breakdown. What are your thoughts on this? I don't think its too user intrusive.

bqpd commented:
<img width="523" alt="screen shot 2016-02-02 at 5 29 03 pm" src="https://cloud.githubusercontent.com/assets/6042395/12766554/844577cc-c9d2-11e5-888f-8cd185368749.png">
<img width="463" alt="screen shot 2016-02-02 at 5 29 11 pm" src="https://cloud.githubusercontent.com/assets/6042395/12766555/844e0248-c9d2-11e5-98f0-3c70008abd9f.png">

whoburg commented:
I still feel like this is dying to be a ConstraintSet that can be included as a model building block. The drawing functionality is a natural use for `process_result`.

In the current form I'm worried that models will get out of sync from their corresponding Breakdown.

I think all the work done here on drawing will be super useful, and I don't think defining the recursive class will be too bad. I can take a shot at this at some point if it's confusing what I'm suggesting here.

mayork commented:
What do you want the input to the building block constraint set to be?

mayork commented:
Also, I can adjust line styles etc. is there a preference on what they should look like?

whoburg commented:
1. I would think there would be two main inputs:

```python
Arguments
---------
root (varkey):
    top-level variable to break down
leaves (iterable):
    list of (variables or other breakdowns) that sum to form root
```

The resulting constraint pseudocode is something like

```python
class Breakdown(Model):
    def __init__(self, **kwargs):
        constraints = [root >= sum(getattr(x, "root", x) for x in leaves)]
        constraints.extend([x for x in leaves if hasattr(x, "root")])  # add sub-breakdown constraints
        self.root = root
        Model.__init__(root, constraints, **kwargs)
```
... and then the only hard part is drawing. Unless I'm missing something hard?

On style, I think we should try to make them look like the latex examples posted above.

mayork commented:
that should be a pretty easy refactor on making the breakdown constraint set the way you want...I think most of the current code can actually still be used.

As far as the drawings do we want to actually make them tin latex? I'm rather amateur with latex so I'm not sure how hard it would be to make the switch.  I know @bqpd suggested svg, but there might be advantages to latex (namely simpler printing of math symbols)

bqpd commented:
Yeah you probably do want to make them in latex, though might be worth
doing svg as well for faster interactive display. Latex is just too slow
for interactive use.

On Sep 13, 2016 6:25 PM, "mayork" <notifications@github.com> wrote:

that should be a pretty easy refactor on making the breakdown constraint
set the way you want...I think most of the current code can actually still
be used.

As far as the drawings do we want to actually make them tin latex? I'm
rather amateur with latex so I'm not sure how hard it would be to make the
switch. I know @bqpd <https://github.com/bqpd> suggested svg, but there
might be advantages to latex (namely simpler printing of math symbols)

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/hoburg/gpkit/pull/522#issuecomment-246845317>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/ABagGL38VWn3FpxyBFSjJ6a9fB2rBlFNks5qpyK_gaJpZM4HiJyG>
.

whoburg commented:
yeah tough call there. Possibly useful: http://tex.stackexchange.com/questions/51757/how-can-i-use-tikz-to-make-standalone-svg-graphics
